2018 Robot/src/org/usfirst/frc/team4999/utils/PIDFactory.java

```diff
@@ -0,0 +1,393 @@
+			this.left = left;
+			this.right = right;
+		}
+		
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			sourcetype = pidSource;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return sourcetype;
+		}
+
+		@Override
+		public double pidGet() {
+			switch(sourcetype) {
+			case kRate:
+				return (left.getRate() + right.getRate()) / 2;
+			case kDisplacement:
+			default:
+				return (left.getDistance() + right.getDistance()) / 2;
+			}
+		}
+		
+	}
+	
+	static class GyroPitch implements PIDSource {
+		PIDSourceType type = PIDSourceType.kDisplacement;
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			type = pidSource;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return type;
+		}
+
+		@Override
+		public double pidGet() {
+			double angle = RobotMap.gyro.getPitch();
+			if(RobotMap.gyro.currentSensor() == Sensor.ADIS)
+				angle = 0;
+			double tiltRange = MoPrefs.getTiltRange();
+			if(angle > tiltRange)
+				return angle - tiltRange;
+			else if(angle < -tiltRange)
+				return angle + tiltRange;
+			else
+				return 0;
+		}
+		
+	}
+	
+	public static MomentumPID getMovePID() {
+		MomentumPID ret;
+		String path = BASE + MOVE_FILE;
+		//PIDSource source = new AverageEncoder(RobotMap.leftDriveEncoder, RobotMap.rightDriveEncoder);
+		PIDSource source = RobotMap.leftDriveEncoder;
+		source.setPIDSourceType(PIDSourceType.kDisplacement);
+		if(checkFile(path)) {
+			Properties props = openFile(path);
+			double p = Double.parseDouble(props.getProperty("P", DEFAULT_P));
+			double i = Double.parseDouble(props.getProperty("I", DEFAULT_I));
+			double d = Double.parseDouble(props.getProperty("D", DEFAULT_D));
+			double errZone = Double.parseDouble(props.getProperty("Error Zone", DEFAULT_ERR_ZONE));
+			double targetZone = Double.parseDouble(props.getProperty("Target Zone", DEFAULT_TARGET_ZONE));
+			double targetTime = Double.parseDouble(props.getProperty("Target Time", DEFAULT_TARGET_TIME));
+			ret = new MomentumPID("MovePID",p,i,d,errZone,targetZone,source, null);
+			ret.setTargetTime(targetTime);
+		} else {
+			double p = Double.parseDouble(DEFAULT_P);
+			double i = Double.parseDouble(DEFAULT_I);
+			double d = Double.parseDouble(DEFAULT_D);
+			double errZone = Double.parseDouble(DEFAULT_ERR_ZONE);
+			double targetZone = Double.parseDouble(DEFAULT_TARGET_ZONE);
+			double targetTime = Double.parseDouble(DEFAULT_TARGET_TIME);
+			ret = new MomentumPID("MovePID",p,i,d,errZone,targetZone,source, null);
+			ret.setTargetTime(targetTime);
+		}
+		ret.setListener(()->saveMovePID(ret));
+		addToCalculator(ret);
+		return ret;
+	}
+	
+	public static void saveMovePID(MomentumPID pid) {
+		savePID(pid, BASE+MOVE_FILE);
+	}
+	public static MomentumPID getTurnPID() {
+		MomentumPID ret;
+		String path = BASE + TURN_FILE;
+		PIDSource source = new GyroFusion();
+		source.setPIDSourceType(PIDSourceType.kDisplacement);
+		if(checkFile(path)) {
```

Comment:
Lots of duplication here.  It seems like these get*PID methods can be refactored to share the bulk of their code.